kep/numa-nodes-scoring/README.md

```diff
@@ -0,0 +1,280 @@
+in latency-critical execution and high-throughput applications.
+Topology Manager assigns resources from least amount of NUMA nodes but the scheduler is unaware
+of different NUMA topologies. The best case scenario would be to schedule pod on the node that
+can satisfy resource requirements from least amount of NUMA nodes to minimize latency.
+
+## Goals
+
+- Make better scheduling decisions that take NUMA topology into consideration.
+
+## Non-Goals
+
+- Change the PodSpec to allow requesting a maximum amount of NUMA nodes required to satisfy resource requirements.
+
+# Proposal
+
+A new ScoringStrategy `LeastNUMANodes` would check how many NUMA nodes are required to run given pod. It will use [CRD][1]
+to get available resources on the worker nodes and identify which topology policy is enabled, the same as other ScoringStrategies.
+
+For now available ScoringStrategies are running only for `single-numa-node` TopologyManager policy.
+`LeastNUMANodes` strategy can score nodes running all TopologyManager policies.
+
+## ScoringStrategy implementation details
+
+Since topology of the node is stored in the CR, kube-scheduler should be subscribed for updates of appropriate CRD type.
+Same mechanism is used by other ScoringStrategies for `NodeResourceTopologyMatch` plugin.
+
+### Description of calculating required NUMA nodes
+
+To calculate how many NUMA nodes are required to run given container/pod the plugin will generate combinations of all NUMA nodes for given bitmask and see if there are enough resources in specific combination.
+It's important to notice the we need to mimic how Topology Manger chooses which bitmask is the most suitable if there are multiple combinations of NUMA nodes that can satisfy the request.
+The bitmask with less bits set or if the same number of bits is set then the mask with more lower-numbered bits set wins out([link](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/bitmask/bitmask.go#L143)).
+The [Combination][2] function can used for that since it generate combinations from the lowest values.
+
+```go
+func getRequiredNUMANodes(logKey string, qos v1.PodQOSClass, numaNodes noderesourcetopology.NUMANodeList, resources v1.ResourceList, nodeName string) (bitmask.BitMask, int) {
+	combinationBitmask := bitmask.NewEmptyBitMask()
+    // if the node only has 1 NUMA node return [0], 1
+	if len(numaNodes) == 1 {
+		combinationBitmask.Add(0)
+		return combinationBitmask, 1
+	}
+
+	// we will generate combination of numa nodes from len = 1 to the number of numa nodes present on the machine
+	for i := 1; i <= len(numaNodes); i++ {
+		// generate combinations of len i
+		numaNodesCombination := combin.Combinations(len(numaNodes), i)
+		// iterate over combinations for given i
+		for _, combination := range numaNodesCombination {
+			// accumulate resources for given combination
+			combinationResources := combineResources(numaNodes, combination)
```

Comment:
Can you please clarify what `combineResources` function does here? Does it return the total amount of resources  present on the evaluated NUMA Node combination?

Reply:
Yes, it returns the sum of all Resources in given NUMA Node combination. I'll add it to proposal.

Reply:
Added.